### PR TITLE
feat(engine): add ShadowRealm behind --unsafe-shadowrealm

### LINKED ISSUE #777: Implement ShadowRealm behind an explicit unsafe flag
## Summary

Implement the TC39 ShadowRealm proposal as an opt-in global built-in, exposed only when a new explicit flag is enabled.

Recommended flag name: `--unsafe-shadowrealm`, with matching config key `"unsafe-shadowrealm": true`.

The proposal specifies `ShadowRealm` as a global ECMAScript built-in constructor, not as a test262-only host hook. Because `ShadowRealm.prototype.evaluate()` performs dynamic source evaluation, GocciaScript should keep it unavailable by default and require a dedicated unsafe capability flag before installing `globalThis.ShadowRealm`.

## Why

The pinned built-ins test262 run currently has 64 ShadowRealm failures:

- `built-ins/ShadowRealm`: 64
- `built-ins/ShadowRealm/prototype`: 51
- `built-ins/ShadowRealm/prototype/evaluate`: 37
- exact message cluster: `ReferenceError: ShadowRealm is not defined`: 57

TC39 proposal status:

- Proposal: ShadowRealm
- Stage: 2.7
- Spec: ECMA-262 proposal
- test262 flags: `ShadowRealm`, `realms-tests`
- Proposal index SHA observed via TC39 MCP: `7510bde4071c3251114d1e7a244187c7c54da214`

The draft says `%ShadowRealm%` has global name `ShadowRealm`, and section 3.2 says the constructor is the initial value of the `"ShadowRealm"` property of the global object:
https://tc39.es/proposal-shadowrealm/

GocciaScript's project policy keeps dynamic code generation out of normal runtimes by default. `eval()` is private to the test262 host, and `Function()` already requires `--unsafe-function-constructor`. ShadowRealm should follow that explicit-capability pattern, but with its own flag so hosts can opt into this proposal independently.

## Current behavior

`ShadowRealm` is not installed.

test262 failures report `ReferenceError: ShadowRealm is not defined` for most ShadowRealm tests.

GocciaScript already has private realm/eval machinery for conformance:

- `$262.evalScript()`
- `$262.createRealm()`
- `GocciaScriptLoaderBare --test262-host`

Those hooks are not a public ShadowRealm implementation and should remain conformance host capabilities.

## Expected behavior

When `--unsafe-shadowrealm` is not enabled:

- `globalThis.ShadowRealm` is absent or unavailable, preserving the current no-dynamic-eval default.
- Normal runtime profiles do not gain a new source-evaluation capability.

When `--unsafe-shadowrealm` is enabled:

- `ShadowRealm` is installed as a global constructor.
- `new ShadowRealm()` creates a fresh realm with its own global object and intrinsics.
- `ShadowRealm()` without `new` throws `TypeError`.
- `ShadowRealm.prototype.evaluate(sourceText)` evaluates script source in the target ShadowRealm.
- Non-primitive, non-callable cross-realm results throw according to the proposal's callable-boundary rules.
- Callable results are exposed through wrapped function exotic behavior or the closest GocciaScript equivalent.
- `ShadowRealm.prototype[Symbol.toStringTag]` is `"ShadowRealm"`.
- test262 ShadowRealm and realms-tests cases can be enabled through the new flag.

## Scope notes

Likely implementation areas:

- CLI/config flag wiring near `--unsafe-function-constructor`
- `GocciaScriptLoaderBare` test262 host integration
- `TGocciaEngine.RegisterBuiltIns` global registration
- realm creation/evaluation machinery currently used by `$262.createRealm()` and `$262.evalScript()`
- test262 runner feature-flag handling for `ShadowRealm` / `realms-tests`

Non-goals for this issue:

- Do not expose ShadowRealm by default.
- Do not fold this into `--unsafe-function-constructor`; ShadowRealm needs a separate capability gate.
- Do not expose `$262` host hooks as public APIs.
- Do not implement broader Compartments or host virtualization APIs.
- Do not claim ShadowRealm is a security boundary; the proposal itself says it does not provide availability protection and is not a full confidentiality mechanism.

### PR DESCRIPTION
**Pull request by frostney**
## Summary

Implements the [TC39 ShadowRealm proposal](https://github.com/tc39/proposal-shadowrealm) (stage 2.7) as an opt-in core language built-in, gated by a new `--unsafe-shadowrealm` flag / `"unsafe-shadowrealm": true` config key. It is off by default because `ShadowRealm.prototype.evaluate` performs dynamic source evaluation, which GocciaScript keeps out of normal runtimes — the same explicit-capability pattern as `--unsafe-function-constructor`.

**What's implemented**

- `globalThis.ShadowRealm` constructor — engine-owned, installed only when the flag is set (genuinely absent otherwise). `new` required (`ShadowRealm()` throws `TypeError`); `length` 0, `name` `"ShadowRealm"`, `[[Prototype]]` = `Function.prototype`; `ShadowRealm.prototype[Symbol.toStringTag]` = `"ShadowRealm"`.
- `ShadowRealm.prototype.evaluate(sourceText)` — runs source in a fresh child realm (its own intrinsics + global object) using eval-style declaration instantiation: `var`/`function` persist on the realm's `globalThis`, `let`/`const` are per-call. Parse failures throw a caller-realm `SyntaxError`; runtime errors throw a caller-realm `TypeError`. Nested realms are supported.
- Cross-realm callable boundary as wrapped function exotic objects: primitives pass through, non-callable objects throw `TypeError`, callables are wrapped (name/length copied, arguments/results marshaled across realms, thrown errors re-wrapped into the caller realm).
- test262 `ShadowRealm` / `realms-tests` features map to `--unsafe-shadowrealm` in the conformance runner.

**Key implementation notes**

- Engine-owned rather than a runtime extension, because the runtime-less Bare Script Loader used by test262 has no runtime layer; each instance owns a child `TGocciaEngine` freed at the creating engine's teardown (a GC sweep can't safely free a child engine from a collected instance). See [ADR 0073](docs/adr/0073-opt-in-shadowrealm.md).
- `InvokeCallable` now routes callable non-function/class values (e.g. callable proxies) through the proxy apply hook so wrapped callable proxies work — previously they raised "object is not a function".

**Deferred / out of scope (follow-ups)**

- `ShadowRealm.prototype.importValue` (module import in the child realm) — a separate proposal method, not in this issue's expected behavior (~12 test262 cases).
- Child realms intentionally do not expose `eval`, consistent with the no-eval default.
- Cross-realm `Symbol.for` registry sharing — a pre-existing per-realm limitation also visible via `$262.createRealm`.
- Strict-mode early errors (e.g. assigning to `arguments`) and native-function `length` deletability are pre-existing engine limitations surfaced by a few test262 cases.
- `FWrappedHosts` retains wrapped-function state for the realm's lifetime (memory bounded by engine lifetime; candidate for a future GC-owned wrapper).

Closes #777

## Testing

- [x] Verified no regressions and confirmed the new feature or bugfix in end-to-end JavaScript/TypeScript tests — full JS suite **10905/10905** passing in **both** interpreter and bytecode modes (after merging `origin/main`); 35 new `tests/built-ins/ShadowRealm/**` tests. test262 `built-ins/ShadowRealm` at the pinned SHA: **48/64** (was 0/64), **0** wrapper-infra failures; the remaining 16 are the deferred `importValue` (12) and the pre-existing-limitation edge cases (4) noted above. `./format.pas --check` clean.
- [x] Updated documentation — `docs/built-ins.md`, `docs/language-tables.md`, `docs/build-system.md`, and ADR 0073.
- [ ] Optional: Verified no regressions and confirmed the new feature or bugfix in native Pascal tests (if AST, scope, evaluator, or value types changed) — n/a; additive built-in, no AST/scope/value-type changes.
- [ ] Optional: Verified no benchmark regressions or confirmed benchmark coverage for the change — n/a.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

File: source/units/Goccia.Builtins.GlobalShadowRealm.pas [hunk 1]
```diff
@@ -0,0 +1,577 @@
...
+  DestScope := FEngine.ActivateRealmExecutionContext;
+  try
+    Wrapped := TGocciaNativeFunctionValue.CreateWithoutPrototype(
+      WrappedHost.Invoke, TargetName, TargetLength);
+  finally
+    DestScope.Free;
+  end;
+  // Keep the wrapped target reachable for the GC through the wrapper function.
+  Wrapped.CapturedRoot := AValue;
+  Result := Wrapped;
```
[the remark this answers is not in the document]
Comment (frostney) [reply source/units/Goccia.Builtins.GlobalShadowRealm.pas]: Fixed in 28812d19. `WrapIncoming` now temp-roots the target across the `length`/`name` reads (which can run a proxy get-trap — a GC safe point) and the wrapper allocation until `CapturedRoot` is set, and `Invoke` roots the wrapped receiver and each wrapped argument during the build loop until `InvokeCallable` takes over their rooting — mirroring `InvokeCallable`'s existing input-rooting pattern.

(The common plain-function path was already safe, since GocciaScript's GC is cooperative: `RegisterObject` only bumps counters and collection runs at explicit safe points, not on allocation. But the proxy-trap-during-wrap edge is real, so the rooting is warranted.)

File: docs/build-system.md [hunk 2]
```diff
@@ -290,6 +290,7 @@ Relative paths are resolved against the current working directory. A missing fil
   "experimental-js-module-source": true,
   "strict-types": true,
   "unsafe-ffi": true,
+  "unsafe-shadowrealm": true,
```
[the remark this answers is not in the document]
Comment (frostney) [reply docs/build-system.md:293]: Good catch on the divergence — I've left the TOML example unchanged here, though. `docs/build-system.md` is already at the project's hard 800-line `check-doc-length` limit (after merging `origin/main`), so adding `unsafe-shadowrealm = true` pushes it to 801 and fails the pre-commit gate. The JSON config example above already documents the flag, so rather than bump the file's length budget (or trim unrelated content) for a one-line cosmetic mirror, I kept it under the limit. Happy to add it behind a `<!-- doc-length-limit: ... -->` override if you'd prefer the two examples mirror exactly.
